### ISSUE TITLE
Per-call model.Train (tape path) with default Noam scheduler keeps LR frozen at warmup-step-1 -> from-scratch Transformer training stalls (regression 0.207.0-fix1380 -> 0.207.x)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1470]
## Summary
A from-scratch Transformer trained via the **per-call minibatch pattern** (`MaxIterations=1` + an external epoch loop calling `model.Train(xBatch, yBatch)` repeatedly) **stalls** on AiDotNet 0.207.x / Tensors 0.86.x but **converges** on 0.207.0-fix1380 / Tensors 0.81.8. Controlled A/B: identical code, data, seed, config — only the package pins differ.

| pin | epoch-1 loss | epoch-8 loss | held-out top-1 |
|---|---|---|---|
| **0.207.0-fix1380 / Tensors 0.81.8** | 9.05 (9.61→8.19 within ep1) | **6.6** | **11.07%** ✅ |
| **0.207.7 / Tensors 0.86.3** | 9.61 (flat) | 8.85 | 1.02% ❌ |
| **0.207.9 / Tensors 0.86.6 (latest)** | 9.61 (flat) | 8.85 | 1.02% ❌ |

(V=15001 BPE, d=128, h=8, L=2, ctx=64, 300K WT103 tokens, Adam lr=1e-3, CPU.) The new-pin loss barely leaves uniform (ln 15001 ≈ 9.6) ⇒ the **effective LR is ~the warmup-step-1 value (~1e-7)**, not ramping.

## Root cause (narrowed)
- The model is `new Transformer(arch, loss)` with **no optimizer**, so it uses `CreateDefaultVaswaniOptimizer()` = Adam + **`NoamSchedule`** with **`SchedulerStepMode.StepPerBatch`**. This default is **identical** in 0.207.0-fix1380 and 0.207.x.
- `Train` routes the tape path to `TrainWithTape(..., optimizer:null)` → `GetOrCreateBaseOptimizer()`; `OnBatchEnd()` is called per batch and **correctly** does `StepPerBatch ⇒ StepScheduler()`.
- **Yet the LR stays pinned at the warmup-step-1 value** across thousands of per-call `Train` invocations on 0.207.x. On 0.207.0-fix1380 the same path ramps the Noam LR and converges.
- This points at the **#1413 optimizer consolidation** ("consolidate flat-vector UpdateSolution into Step delegation across all 19 optimizers"): the per-call `TrainWithTape` no longer accumulates the scheduler/optimizer step state across calls (suspected: `GetOrCreateBaseOptimizer` instance/scheduler-state, or the consolidated `Step(TapeStepContext)` not reading the scheduler's refreshed `CurrentLearningRate`).

## Repro
Per-call minibatch loop (the standard "build [B,V] on the fly to avoid the [N,V] one-hot OOM at V=50257" pattern):
```csharp
var model = new Transformer<float>(arch, lossFunction: new CategoricalCrossEntropyLoss<float>()); // default Noam opt
for (int ep = 0; ep < epochs; ep++)
  foreach (var (xB, yB) in minibatches)   // MaxIterations=1
      model.Train(xB, yB);                // tape path; LR stays at warmup-step-1 on 0.207.x
```

## Confirmed workaround
Pass a **fixed-LR Adam (no scheduler)** to the ctor — converges on the latest packages:
```csharp
var opt = new AdamOptimizer<float,Tensor<float>,Tensor<float>>(null,
    new AdamOptimizerOptions<float,Tensor<float>,Tensor<float>>{ InitialLearningRate = 1e-3 });
var model = new Transformer<float>(arch, loss, optimizer: opt);  // ⇒ loss 9.64→7.54 ep1, 4.77% @3ep
```

## Proposed fix
Ensure the **per-call `TrainWithTape` path accumulates scheduler state across calls** so a `StepPerBatch` schedule (e.g. the default Noam) ramps as documented — i.e. `GetOrCreateBaseOptimizer()` must return the persisted instance and the consolidated `Step(TapeStepContext)` must use the scheduler's current `CurrentLearningRate` (refreshed by `OnBatchEnd`). A regression-guard unit test (small Transformer, per-call `Train` with default Noam, assert LR ramps past warmup-step-1 / loss drops below a threshold over N batches) would lock it.

## Environment
gfx1012/RDNA1, CPU engine, net10.0. Diagnosis from a controlled A/B in a downstream consumer (HarmonicEngine).

**Comment by ooples** [comment]
### Root cause confirmed — constant-rate fused-training kernel freezes the Noam ramp

Traced the full data flow. The eager training path is correct: per-call `Train` → `TrainWithTape` → `opt.Step(context)` (reads `CurrentLearningRate`) → `StepSchedulerIfSupported` → `OnBatchEnd` → `StepScheduler` → `SetLearningRate(scheduler.Step())`. The optimizer instance persists across calls (`GetOrCreateBaseOptimizer` caches via `??=`; the Transformer ctor wires the Noam optimizer as `_baseTrainOptimizer`), so the scheduler step **does** accumulate on the eager path.

The freeze is the **compiled fused-training path**, which bakes a *constant* learning rate. On `master` (0.207.x) a default Transformer's Adam + `NoamSchedule` (`StepPerBatch`) optimizer gets committed to that fused path on the first successful fused step (`_fusedTrainingCommitted = true`), after which the LR is pinned at whatever it was at commit time (the warmup-step-1 value) — the per-step Noam ramp can't be reproduced by a constant-rate kernel. That matches the observed `9.61 (flat)` loss and effective LR ≈ 1e-7.

### Fix (on the `perf/fused-training-gate` branch / PR #1469)

The branch adds the `IFusedOptimizerSpec` gating: `AdamOptimizer.TryGetFusedOptimizerConfig` → `TryGetFusedLrSchedule` returns **false** for any schedule the fused kernel can't reproduce (Noam hits the `default` case). `NeuralNetworkBase.TryMapToFusedOptimizerConfig` delegates to that spec, so an Adam + Noam model now **declines the fused path and falls back to the eager `OnBatchEnd`/`StepScheduler` path that ramps the LR**. Cosine/Exponential schedules are mapped into the fused kernel and still take the fast path. The confirmed workaround (explicit fixed-LR Adam) keeps working because a no-scheduler Adam maps cleanly to a constant fused rate.

### Regression guards added (commit `7310df990`)

1. `AdamWithNoamSchedule_DoesNotMapToConstantRateFusedConfig` — deterministic unit test of the exact fix seam: `Adam+Noam → TryGetFusedOptimizerConfig == false` (forces eager), no-scheduler `Adam == true` (fast path preserved).
2. `Transformer_PerCallTrain_DefaultNoam_RampsLearningRateAcrossCalls` — end-to-end smoke: a default-optimizer Transformer trained via repeated per-call `Train` accumulates one scheduler step per call and ramps the Noam LR above its warmup-step-1 value.

Both pass on the branch. **Leaving this open** — the fix rides PR #1469 and isn't on `master`/a release yet; it can be closed when #1469 merges.

**Comment by ooples** [comment]
### Correction — the real fix: Noam on the fused path (not eager fallback)

The eager-fallback approach above is a workaround, not the solution — it forces Adam+Noam Transformers off the fast path. The correct fix mirrors PyTorch's `fused=True` optimizers, which take `lr` as a per-step scalar while the schedule runs alongside the kernel.

The fused training plan **already** evaluates `LrSchedule.GetLr(step)` every optimizer step (`CompiledTrainingPlan`: `_optimizerStep++; lr = GetLr(_optimizerStep)`), so any schedule shape gets true per-step adaptive LR on the fused path. Cosine/Exponential/OneCycle/LinearWarmupCosine were already supported — **Noam was just the missing shape**, which is the only reason Transformers fell back to eager.

**Real fix (Tensors): [AiDotNet.Tensors PR #504](https://github.com/ooples/AiDotNet.Tensors/pull/504)** — adds `LrSchedule.Noam(modelDimension, warmupSteps, factor)` (`factor·d^(-0.5)·min(t^(-0.5), t·warmup^(-1.5))`). Its `t = step` (1-based) matches the fused plan's `_optimizerStep` and the eager `NoamSchedule`, so the fused and eager LR sequences are bit-identical (`lr(t=N)` on batch N). 5 unit tests passing.

**AiDotNet follow-up (after the Tensors release):** map `NoamSchedule → LrSchedule.Noam(...)` in `GradientBasedOptimizerBase.TryGetFusedLrSchedule` (replacing the `default → false` eager fallback), bump the Tensors package, and flip the #1470 regression guard to assert Adam+Noam now **engages** the fused path (instead of declining it). Tracked on `perf/fused-training-gate` / PR #1469.

This keeps the Transformer default on the fast path **with** a correct warmup ramp — no forced slow path.

**Comment by ooples** [comment]
### Wired up + verified — Adam+Noam now trains on the fused fast path

AiDotNet.Tensors **0.88.0** (with `LrSchedule.Noam`, Tensors #504) is now consumed, and the AiDotNet side is wired on `perf/fused-training-gate` / PR #1469 (commit `53f58b4e6`):

- Bumped `AiDotNet.Tensors` 0.86.6 → **0.88.0** (Native packages in lockstep).
- `GradientBasedOptimizerBase.TryGetFusedLrSchedule`: maps `NoamSchedule → LrSchedule.Noam(modelDim, warmup, factor)` (replaces the `default → false` eager fallback). Both use `t = step` (1-based) so the fused LR sequence is bit-identical to the eager `NoamSchedule`.
- Added `NoamSchedule.Factor` getter for a faithful mapping.

**Integration tests (3, all green on the live CUDA box):**
1. **Gating** — Adam+Noam now maps to a fused config (no eager fallback); the mapped schedule ramps **4000×** over warmup and hits the exact paper peak.
2. **Equivalence** — fused `GetLr(N)` == eager `lr(t=N)` for 3× warmup steps.
3. **End-to-end** — a default-Noam Transformer trained via the per-call minibatch pattern **engages the fused path on every step (3200/3200)** and converges to **PPL 5.06, top-1 7/8** (`avgNll 1.62 < ln(V) 2.08`) — proving the Noam LR ramps on the fused path instead of freezing.

This is the real fix the issue called for: the Transformer default stays on the fused fast path **with** a correct warmup ramp.

Note: the pre-existing ModelFamily `TableTransformer`/`TabTransformer`/`DecisionTransformer` failures on this branch reproduce identically at the 0.86.6 baseline, don't use Noam, and are unrelated to this change.

Closes when PR #1469 merges.
